### PR TITLE
async scheduler fix  _substitute_placeholder_token_fn  bug

### PR DESCRIPTION
**Pull request by cychiuak**
# Description

This commit aims to fix a bug in _substitute_placeholder_token_fn where in an edge case unexpected behaviour occured. This commits have verified to retain the performance gain of async scheduler.

The edge case is that when we aim to update the last element in input_ids and padding is required, the last element is updated to its original value instead of the new value:
our function:
```
def _substitute_placeholder_token(
        input_ids: jax.Array, token_in_tpu_cur_input_indices: jax.Array,
        token_in_tpu_pre_next_tokens_indices: jax.Array,
        next_tokens: jax.Array, placeholder_num: int):
        
    # updates the input_ids for all placeholders.
    mask = jnp.arange(input_ids.shape[0]) < placeholder_num
    new_token_values = next_tokens[token_in_tpu_pre_next_tokens_indices]
    original_values = input_ids[token_in_tpu_cur_input_indices]
    update_values = jnp.where(mask, new_token_values, original_values)
    return input_ids.at[token_in_tpu_cur_input_indices].set(update_values)
```
test:
```
import jax
import jax.numpy as jnp

# inputs to the substitute_placeholder_token_fn
placeholder_num = 2
input_ids =  jnp.array([    0,     127,   264, 32794,   911,   279,  5112,   315,   264,
    3283,   518,  3729,    13,  9645,   432,   0], dtype=jnp.int32)
token_in_tpu_pre_next_tokens_indices = jnp.array([ 0,  2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], dtype=jnp.int32)
token_in_tpu_cur_input_indices = jnp.array([ 0,  15, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], dtype=jnp.int32)  # Original padding
next_tokens = jnp.array([1265,  264,  576, 1265, 1265, 1265, 1265, 1265], dtype=jnp.int32)

out = _substitute_placeholder_token(input_ids, token_in_tpu_cur_input_indices, token_in_tpu_pre_next_tokens_indices, next_tokens, placeholder_num)

print("input_ids is \n",input_ids)
print("update_values is \n",update_values)
print("output is \n",out)
print("we expect output is [ 1265   127   264 32794   911   279  5112   315   264  3283   518  3729 \n    13  9645   432     576]")
```
output:
```
input_ids is 
 [    0   127   264 32794   911   279  5112   315   264  3283   518  3729
    13  9645   432     0]
update_values is 
 [1265  576    0    0    0    0    0    0    0    0    0    0    0    0
    0    0]
output is 
 [ 1265   127   264 32794   911   279  5112   315   264  3283   518  3729
    13  9645   432     0]
we expect output is [ 1265   127   264 32794   911   279  5112   315   264  3283   518  3729 
    13  9645   432     576]
```

Why the original padding has bug but works in vLLM? 
Such a scenario will not happen in vLLM as only decode tokens need to be updated and decode tokens always scheduled before prefill tokens. Therefore, there won't be such a scenario that we need a padding when we update the last token.

Why we try to fix this bug? 
Although such a scenario is unlikely to happen in vLLM, it is best to eliminate any potential risks.
    
Why the new padding on token_in_tpu_cur_input_indices fixed the bug? The padded parts, which we pad with indices that does not need any updates, will just be updating with their original values.

# Tests

We test this fix with the same edge cases and it returns the correct output. We also verified its correctness by running existing pytest on async scheduler.
Result on new padding:
```
import jax
import jax.numpy as jnp

# inputs to the substitute_placeholder_token_fn
placeholder_num = 2
input_ids =  jnp.array([    0,     127,   264, 32794,   911,   279,  5112,   315,   264,
    3283,   518,  3729,    13,  9645,   432,   0], dtype=jnp.int32)
token_in_tpu_pre_next_tokens_indices = jnp.array([ 0,  2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], dtype=jnp.int32)
token_in_tpu_cur_input_indices = jnp.array([ 0,  15, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ,11, 12, 13, 14], dtype=jnp.int32)   # New padding
next_tokens = jnp.array([1265,  264,  576, 1265, 1265, 1265, 1265, 1265], dtype=jnp.int32)

out = _substitute_placeholder_token(input_ids, token_in_tpu_cur_input_indices, token_in_tpu_pre_next_tokens_indices, next_tokens, placeholder_num)
print("input_ids is \n",input_ids)
print("update_values is \n",update_values)
print("output is \n",out)
```
output:
```
input_ids is 
 [    0   127   264 32794   911   279  5112   315   264  3283   518  3729
    13  9645   432     0]
token_in_tpu_cur_input_indices is  [ 0 15  1  2  3  4  5  6  7  8  9 10 11 12 13 14]
update_values is 
 [ 1265   576   127   264 32794   911   279  5112   315   264  3283   518
  3729    13  9645   432]
output is 
 [ 1265   127   264 32794   911   279  5112   315   264  3283   518  3729
    13  9645   432   576]
 ```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/runner/tpu_jax_runner.py [hunk 1]
```diff
@@ -937,10 +944,14 @@ def _prepare_inputs(self, scheduler_output: "VllmSchedulerOutput"):
...
-            padded_token_in_tpu_cur_input_indices = np.pad(
-                token_in_tpu_cur_input_indices, (0, idx_pad_len),
-                mode='constant',
-                constant_values=-1)
+
+            # Pad according to the instructions written inside self._substitute_placeholder_token_fn
+            full_range = np.arange(0, len(input_ids))
+            missing_values = np.setdiff1d(full_range,
+                                          token_in_tpu_cur_input_indices)
+            padded_token_in_tpu_cur_input_indices = np.concatenate(
```
Comment (py4) [inline tpu_inference/runner/tpu_jax_runner.py:1026]: Shouldn't this and other inputs to `_substitute_placeholder_token_fn` first put on device then passed to the function? That function is a jitted function and expects args to be on device already

Comment (cychiuak) [reply tpu_inference/runner/tpu_jax_runner.py:1026]: Yes the inputs are put into device at [previous steps](https://github.com/vllm-project/tpu-inference/blob/main/tpu_inference/runner/tpu_jax_runner.py#L931C1-L934C74)

Comment (cychiuak) [reply tpu_inference/runner/tpu_jax_runner.py:1026]: As for token_in_tpu_cur_input_indices and padded_token_in_tpu_cur_input_indices, according to my understanding, they are converted to JAX  arrays by implicit conversion when we pass np arrays into jax.jit functions. Therefore, they are all in device when we pass in the inputs.

Comment (py4) [reply tpu_inference/runner/tpu_jax_runner.py:1026]: @cychiuak This is not a good practice i think. The convention in the codebase is that we explicitly move data to device before calling the jitted function. That way is readable i think. Otherwise it's hard to understand which args are on device, which args on cpu and makes it hard to reason about a function
